File: 105. Construct Binary Tree from Preorder and Inorder Traversal.go
func buildTree(preorder []int, inorder []int) *TreeNode {
	mp := make(map[int]int)
	for i, val := range inorder{
		mp[val] = i
	}
	ind := 0
    
	var dfs func(l, r int) *TreeNode
	dfs = func(l, r int) *TreeNode{
		if(l > r){
			return nil
		}
		rootVal := preorder[ind]
		ind++;
		root := &TreeNode{Val: rootVal}
		mid := mp[rootVal]
		root.Left = dfs(l, mid - 1)
		root.Right = dfs(mid + 1, r)
        return root
	}
	return dfs(0, len(preorder) - 1)
}	

File: 105. Construct Binary Tree from Preorder and Inorder Traversal.py
class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        Map = {}
        for i, a in enumerate(inorder):
            Map[a] = i

        ind = 0
        def dfs(l, r):
            if l > r:
                return None

            nonlocal ind
            root_val = preorder[ind]
            ind += 1
            root = TreeNode(root_val)
            mid = Map[root_val]
            root.left = dfs(l, mid - 1)
            root.right = dfs(mid + 1, r)

            return root

        return dfs(0, len(preorder) - 1)